fix: fix workspace status filter returning more statuses that requested

Emyrk opened the pull request:
Fixes the postgres condition to not make failed workspaces show up with the deleted status filter.
Also fixed the dbfake to match the postgres implementation. It was acting a lot differently.

See https://github.com/coder/coder/issues/7685#issuecomment-1568993952
Fixes: https://github.com/coder/coder/issues/7685

# Testing

Made a unit test to test this logic. Doing it on a real DB is slow, so going to skip in CI :cry:. Took a bit of work to get the test setup.

Emyrk committed:
WIP unit test for workspace statuses

Emyrk committed:
Fix permission issues of dbgen

Emyrk committed:
Write unit test for workspace status

Emyrk committed:
Add unit test for statuses

Emyrk committed:
Import order

Emyrk committed:
Fix test for slow cases

Emyrk committed:
Add unit test to check for filter status

Emyrk committed:
Match dbfake to postgres

Emyrk committed:
Fix test

Emyrk commented on coderd/database/dbgen/generator.go:
```diff
@@ -17,14 +17,25 @@ import (
 // maintained. Any fields omitted will have sensible defaults generated.
 
+// genCtx is to give all generator functions permission if the db is a dbauthz db.
+var genCtx = dbauthz.As(context.Background(), rbac.Subject{
+	ID:     "owner",
+	Roles:  rbac.Roles(must(rbac.RoleNames{rbac.RoleOwner()}.Expand())),
+	Groups: []string{},
+	Scope:  rbac.ExpandableScope(rbac.ScopeAll),
+})
```
This is so dbgen can work on dbauthz wrapped databases for unit testing.

Emyrk commented on coderd/database/queries/workspaces.sql:
```diff
@@ -157,10 +157,12 @@ WHERE
 					latest_build.updated_at - INTERVAL '30 seconds' < NOW() AND
-					latest_build.transition = 'delete'::workspace_transition
+					latest_build.transition = 'delete'::workspace_transition AND
+					-- If the error field is null, the status is 'failed'
+					latest_build.error IS NULL
 
 				WHEN @status = 'deleting' THEN
-					latest_build.completed_at IS NOT NULL AND
+					latest_build.completed_at IS NULL AND
```
This was the only fix needed for prod. All other fixes were just for the dbfake

Emyrk committed:
Linting

Emyrk committed:
Linting

Emyrk committed:
Add dbauthz system comment

spikecurtis commented on coderd/database/queries/workspaces.sql:
```diff
@@ -157,10 +157,12 @@ WHERE
 					latest_build.updated_at - INTERVAL '30 seconds' < NOW() AND
-					latest_build.transition = 'delete'::workspace_transition
+					latest_build.transition = 'delete'::workspace_transition AND
+					-- If the error field is null, the status is 'failed'
```
```suggestion
					-- If the error field is not null, the status is 'failed'
```

spikecurtis commented on coderd/database/queries/workspaces.sql:
```diff
@@ -157,10 +157,12 @@ WHERE
 
 				WHEN @status = 'deleting' THEN
-					latest_build.completed_at IS NOT NULL AND
+					latest_build.completed_at IS NULL AND
```
so, it's really difficult to read this query and convince yourself that these WHERE clauses are 

a) non-overlapping
b) match the golang implementation of the status computation

What do you think about the strategy of introducing a stored procedure that computes the status?  Then the filter logic in this query becomes a trivial WHERE clause, and we can use the DB-computed status in prod, rather than computing the status in the API layer.

Annoyingly, we'd still have to keep the golang implementation around for the dbfake, but they could both be written in an imperative style and so it will be much easier to verify the algorithm matches.

Emyrk replied:
Oh it was a headache. Are we using stored procedures anywhere else, or would this be the first case? I'm open to anything that makes this easier.

Ideally though, it would be in the next PR so this bug is fixed in main. The major contribution from this PR is the test imo.

spikecurtis replied:
We have triggers for deletion of tokens stored in the DB.  This would be the first computed column, though.

Fine to follow up in another PR.

spikecurtis commented on coderd/workspaces_test.go:
```diff
@@ -555,6 +562,201 @@ func TestWorkspaceByOwnerAndName(t *testing.T) {
+	}, database.ProvisionerJob{
+		StartedAt:   sql.NullTime{Time: time.Now().Add(time.Second * -2), Valid: true},
+		CompletedAt: sql.NullTime{Time: time.Now(), Valid: true},
+	}, database.WorkspaceTransitionDelete)
```
these tests are great and all, but they are not combinatorially exhaustive on null vs non-null for all fields.  This is another reason I think a stored procedure for the status is better.

Emyrk replied:
It is definitely not exhaustive. The annoying thing too is when we prepulate this much data (which isn't even that much), the postgres test gets so slow. So I skipped it in CI, meaning this bug could easily pop up again silently :cry:.

I was debating about putting in a batch insert SQL query outside our typical db interface and see if it is faster. If we go the stored procedure route, might be worth investigating how to make the postgres test run in a reasonable amount of time.

spikecurtis replied:
I wonder if the slowness is related to how we have to create the job, then acquire it, then fail/complete it.  Acquiring in particular might be slow because of the locking we do.

In testing we could have some queries that directly insert the job in the state we want it.  Not exposed on the main `Store` interface, but available if you type assert to a test type.

Emyrk committed:
Fix comment

Kira-Pilot commented:
@Emyrk I tried this out locally and failed workspaces do not show up with the deleted status filter - awesome.
I am a bit confused about the status behavior in general, however. If I have a deleted workspace and I filter by `status:deleted`, nothing shows up. This is confusing to me given the ticket description. Am I missing something?
![Screenshot 2023-06-01 at 11 20 03 AM](https://github.com/coder/coder/assets/19142439/376ffb3a-6426-45f4-9754-a5942876cf75)
![Screenshot 2023-06-01 at 11 20 23 AM](https://github.com/coder/coder/assets/19142439/0148233a-20bc-456c-8e0b-3ff2d1f0c9a5)

Emyrk commented:
@Kira-Pilot could it be "deleting"? 
Can you send the "all workspaces" response?

Emyrk commented:
> @Emyrk I tried this out locally and failed workspaces do not show up with the deleted status filter - awesome. I am a bit confused about the status behavior in general, however. If I have a deleted workspace and I filter by `status:deleted`, nothing shows up. This is confusing to me given the ticket description. Am I missing something? ![Screenshot 2023-06-01 at 11 20 03 AM](https://user-images.githubusercontent.com/19142439/242647182-376ffb3a-6426-45f4-9754-a5942876cf75.png) ![Screenshot 2023-06-01 at 11 20 23 AM](https://user-images.githubusercontent.com/19142439/242647203-0148233a-20bc-456c-8e0b-3ff2d1f0c9a5.png)

Ah, so that is a different bug actually.

`deleted` is a special case handled in the sql here: https://github.com/coder/coder/blob/cf35412b77756e08f7da343fce9e7d6108edeaaf/coderd/database/queries/workspaces.sql#L108-L109

What I fixed was the list of statuses. So if you search `status:deleted` nothing shows up as we do not show deleted workspaces to users right now.

So that is expected behavior.

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/workspace_search_failed
